libsolidity/ast/Types.cpp

```diff
@@ -2309,11 +2311,11 @@ u256 StructType::memoryDataSize() const
 	return size;
 }
 
-bigint StructType::storageSizeUpperBound() const
+bigint StructType::sizeUpperBound(DataLocation _loc) const
 {
-	bigint size = 1;
-	for (auto const& member: members(nullptr))
-		size += member.type->storageSizeUpperBound();
+	bigint size = (_loc == DataLocation::Storage) ? 1 : 0; // 1 slot but 0 bytes
```

Comment:
`memoryHeadSize` perhaps?

Reply:
Actually, I'm not sure. This is not really documented in https://solidity.readthedocs.io/en/v0.7.1/internals/layout_in_memory.html

If "head" has the same meaning as in ABI encoded data then `memoryHeadSize()` would be right only for dynamically encoded stuff. Static types are completely stored in the head - so that would include space occupied by struct members.

Reply:
@ekpyron When we estimate a struct size, we add an extra slot for `DataLocation::Storage`. What should we add for other two locations?

```
bigint StructType::sizeUpperBound(DataLocation _loc) const
{
	bigint size = (_loc == DataLocation::Storage) ? 1 : 0; // 1 slot but 0 bytes
	for (MemberList::Member const& member: members(nullptr))
		size += member.type->sizeUpperBound(_loc);
	return size;
}
```

Reply:
I'm a bit out of the loop here I must admit :-).
But for calldata, "dynamically encoded" types are stored using a tail pointer (for which one uint256 of space is required), whereas statically sized types are stored in-place (no additional space is required besides the space for the members itself).
For memory, any reference type is stored behind a pointer, so you'll always need to consider an additional uint256 for any reference type in memory.

That being said, I'm not quite sure where you're currently headed with this ;-). I'll try to take some time to look at the current state of this these days.

Reply:
This is still open.
It should probably be ``1`` for storage, ``isDynamicallyEncoded() ? 1 : 0`` for calldata and ``1`` for memory.

Reply:
"`isDynamicallyEncoded() ? 32 : 0` for calldata and `32`" right? The quirk of this function is that it counts slots in case of storage and bytes in case of calldata and memory.

Reply:
`isDynamicallyEncoded()` causes ICE for
```
pragma experimental ABIEncoderV2;

contract C {
    struct S { function() internal a; }
    function f(S memory) public {}
}
// ----
// TypeError 4103: (103-111): Internal type is not allowed for public or external functions.
```

We can make the error `4103` fatal, or just pragmatically and conservatively fall back to 32 for non-storage types. With a comment. What do you think?